src/Microsoft.NetCore.Analyzers/UnitTests/LeapYear/DoNotIncrementYearValueInDateConstructorTests.cs

```diff
@@ -0,0 +1,240 @@
+            GetCSharpResultAt(24, 27, "orig.Year + 10"),
+            GetCSharpResultAt(25, 38, "orig.Year + 11"),
+            GetCSharpResultAt(26, 27, "orig.Year - 12"));
+        }
+
+        [Fact]
+        public async Task YearIncrementParameterCases()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+    using System;
+    using System.Globalization;
+    using System.Linq;
+
+    namespace LeapYear.UnitTests
+    {
+        public class ClassName
+        {   
+            public void ParameterEdgeCases()
+            {
+                DateTime orig = new DateTime(2000, 1, 2);
+
+                DateTime d1Named = new DateTime(year: orig.Year + 101, month: orig.Month, orig.Day);
+                var d8Named = new DateTime(month: orig.Month, day: orig.Day, hour: 0, minute: 0, second: 1, millisecond: 777, calendar: new GregorianCalendar(), kind: DateTimeKind.Unspecified, year: orig.Year - 801);
+                DateTime d9 = new DateTime(new DateTime(orig.Year + 9, 2, orig.Day).Year, 1, 1);
+                var d21 = new DateTime(orig.Year + 19, 2, 29);
+                var a = 45;
+                DateTime d23 = new DateTime(
+                    a + 19,
+                    orig.Month,
+                    orig.Day);
+                DateTime d23Named = new DateTime(
+                    month: orig.Month,
+                    day: orig.Day,
+                    year: a + 908);
+                DateTime[] d24 = Enumerable
+                    .Range(1, DateTime.DaysInMonth(DateTime.UtcNow.Year, DateTime.UtcNow.Month))
+                    .Select(i => new DateTime(24 + i, i, i, 0, 0, 0, DateTimeKind.Utc)).ToArray();
+                var d31 = new DateTime(orig.Year > 2001 ? orig.Year : orig.Year + 31, orig.Month, orig.Day);
+                var d32 = new DateTime(orig.Year > 2001 ? orig.Year + 32 : orig.Year, orig.Month, orig.Day);
+
+                TimeSpan threeHourTour = new TimeSpan(3, 0, 0);
+                var d10Named = new DateTimeOffset(month: orig.Month, day: orig.Day, hour: 0, minute: 0, second: 0, offset: threeHourTour, year: orig.Year + 101);
+                DateTimeOffset d13 = new DateTimeOffset(new DateTime(orig.Year - 13, 2, orig.Day).Year, 1, 1, 0, 0, 0, new TimeSpan(3, 0, 0));
+            }
+        }
+    }",
+            GetCSharpResultAt(14, 36, "orig.Year + 101"),
+            GetCSharpResultAt(15, 31, "orig.Year - 801"),
+            GetCSharpResultAt(16, 44, "orig.Year + 9"),
+            GetCSharpResultAt(17, 27, "orig.Year + 19"),
+            GetCSharpResultAt(19, 32, "a + 19"),
+            GetCSharpResultAt(23, 37, "a + 908"),
+            GetCSharpResultAt(29, 34, "24 + i"),
+            GetCSharpResultAt(30, 27, "orig.Year + 31"),
+            GetCSharpResultAt(31, 27, "orig.Year + 32"),
+            GetCSharpResultAt(34, 32, "orig.Year + 101"),
+            GetCSharpResultAt(35, 57, "orig.Year - 13"));
+        }
+
+        [Fact]
+        public async Task YearIncrementVariableCases()
```

Comment:
Tagging @dotpaul 

Looking at this test case, it seems like you essentially want to do a version of TaintedDataFlowAnalysis, where an expression is considered tainted if it is a binary operation with one of the operands as a year/month/day of a DateTime variable and other operand is a literal. We already have large number of Taint analysis based rules in the repo, and I think you should build on top of that framework instead of attempting to write your pseudo-dataflow analysis (done in DateKindSemanticModel.cs and DateKindContext.cs at start of this PR).

@dotpaul Your thoughts? If you agree, might be good for you guys to sync in person.

Reply:
Note that if you don't want to handle the DFA cases in this test, and just want the cases in prior tests to work, which might be a reasonable place to start, then you can write an extremely simple IOperation based analyzer (possible 20-30 lines) that can handle those cases. I'd let you guys decide if you want the simpler analyzer that just analyzes arguments or a dataflow analysis based analyzer.

Reply:
As an example, consider the below SQL injection test based on TaintedDataFlowAnalysis: https://github.com/dotnet/roslyn-analyzers/blob/79c2638f54b9e7ef0a3954cc838ec4c13ea63c1f/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Security/ReviewCodeForSqlInjectionVulnerabilitiesTests.cs#L1924-L1962

It has a variable holding tainted data below as the string has a concatenation from an unknown source: https://github.com/dotnet/roslyn-analyzers/blob/79c2638f54b9e7ef0a3954cc838ec4c13ea63c1f/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Security/ReviewCodeForSqlInjectionVulnerabilitiesTests.cs#L1946

This tainted value gets passed to a method call as argument: https://github.com/dotnet/roslyn-analyzers/blob/79c2638f54b9e7ef0a3954cc838ec4c13ea63c1f/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Security/ReviewCodeForSqlInjectionVulnerabilitiesTests.cs#L1948

Then it gets passed to the SqlCommand constructor in the invoked method: https://github.com/dotnet/roslyn-analyzers/blob/79c2638f54b9e7ef0a3954cc838ec4c13ea63c1f/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Security/ReviewCodeForSqlInjectionVulnerabilitiesTests.cs#L1953-L1957

The constructor gets flagged by this unit test.

IMO, this test shows how the analysis required for this analyzer closely matches the one done for SQL injection analyzer and can re-use almost all of the taint analysis dataflow framework.

Reply:
TaintedDataAnalysis would need a bit of modification to treat integer values that get modified as a source.

I think the larger question is, are there other cases would we want to handle that dataflow analysis would help with it? E.g. dunno if something like this would be a case worth trying to find:

```csharp
int y = originalDateTime.Year + 1;
int year = y;
DateTime newDateTime = new DateTime(year, originalDateTime.Month, originalDateTime.Day);
```

(There's also tracking the month and day values for determining if it's February 29, so that's another place where dataflow analysis might be helpful [but even more work].)

But all that being said, dataflow analysis isn't easy, so unless there are harder cases you really want to catch and you have time to work on it, I'd avoid it for now.

Reply:
@dotpaul - are you suggesting that the analyzer be currently written as a simple IOperation analyzer that just flags violations where arithmetic is done in the callsite argument and avoid any DFA based cases? I think that is indeed a reasonable starting point and I am fine with that approach. I am lot more concerned with us attempting to write another pseudo-dataflow analysis solution, which should be avoided.

Reply:
I'm saying I'll defer to @kwilkins as what to do for DFA. :) I dunno if tainted data analysis will meet all the needs. Agree it'd be better to avoid pseudo-DFA. I'd probably consider starting an entirely new analysis myself, given I had the time, but I haven't looked at all the cases.